### PR TITLE
LED Tape

### PR DESCRIPTION
**Pull request by Fasteroid**
I've been procrastinating over opening this PR for far too long, but here it finally is!

preview

LED Tape is a set of beams associated with a controller entity. &nbsp;The set of beams together shares a color defined by whatever you wire to the controller. &nbsp;They glow in the dark and can be paired with wire lights to create very cool-looking effects.

I was able to get performance to acceptable levels with 8 strands of as many as 512 beams each, however, I would love additional feedback and ideas so I can make them even more performance friendly.

### CODE REVIEW

**Review by Denneisk — COMMENTED** [review]
Just minor nitpicks for now. Seems alright.

**Comment by Denneisk** [comment]
I think an "On" input would be nice to have, also. Having it based on the color input is annoying.

File: lua/wire/stools/ledtape.lua
Comment (Denneisk) [inline lua/wire/stools/ledtape.lua:92]: Is there any functional reason it's restricted like this?

Comment (Fasteroid) [reply lua/wire/stools/ledtape.lua:92]: Due to the way the beams are parented to what they attach to, this is unfortunately necessary.  Regular wires (which is what this code is based on) have the same restriction.

File: lua/wire/stools/ledtape.lua
Comment (Fasteroid) [reply lua/wire/stools/ledtape.lua:46]: didn't know that was a thing, will do more updates tomorrow!

File: lua/wire/stools/ledtape.lua
Comment (Denneisk) [inline lua/wire/stools/ledtape.lua:155]: Should abort and notify the user if they have 0 segments.

File: lua/wire/stools/ledtape.lua
Comment (Denneisk) [reply lua/wire/stools/ledtape.lua:92]: Alright, thanks, just didn't know that. Thought it was like ropes.

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:117]: Assuming lua doesn't optimize away creating this empty lambda, branchless isn't gonna help you here. I don't think It'd even help you for lua at all since it's so high level

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:115]: This should not be inside the rendering code

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:110]: Pass number of vertices instead of extravertex?

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:134]: No garry operators.

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:100]: Redundant `, _`

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:101]: No garry operators.

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:209]: You are not verifying that player owns said controller. Or that said controller is even an led tape.

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:299]: I don't think I like this indenting here

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:18]: Address this TODO

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:262]: Inconsistent formatting

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:138]: Why can't we just do all this logic outside the render hook?

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:178]: Either remove and add the hook whenever `#Path >= 2` or simplify to a single bool.

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:192]: Way too much logic here that shouldn't be in a render hook

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:99]: Huge amount of work being done here in a render hook...

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:123]: Repeated use of `#path`, especially in render hook

File: lua/entities/gmod_wire_ledtape.lua
Comment (Vurv78) [inline lua/entities/gmod_wire_ledtape.lua:129]: Calling getColor more times than necessary?

File: lua/wire/stools/ledtape.lua
Comment (Vurv78) [inline lua/wire/stools/ledtape.lua:56]: Length operator in render hook 👎

**Review by Vurv78 — CHANGES_REQUESTED** [review]
Initial reviews

File: lua/entities/gmod_wire_ledtape.lua
Comment (Fasteroid) [reply lua/entities/gmod_wire_ledtape.lua:134]: So no continue?  What specifically is wrong here?

File: lua/entities/gmod_wire_ledtape.lua
Comment (Fasteroid) [reply lua/entities/gmod_wire_ledtape.lua:110]: iirc extravertex changes rendering behavior by duplicating vertices to match the look of normal wires.  I could pass the number of verts and duplicate them ahead of time, but that would increase the number of LocalToWorld calculations for no good reason.

File: lua/wire/stools/ledtape.lua
Comment (Fasteroid) [reply lua/wire/stools/ledtape.lua:56]: Is length operator not cached?

File: lua/entities/gmod_wire_ledtape.lua
Comment (Fasteroid) [reply lua/entities/gmod_wire_ledtape.lua:192]: All of this is based on the rendering code for wires.  Might wanna open an issue for that if this is so bad...

File: lua/entities/gmod_wire_ledtape.lua
Comment (Fasteroid) [reply lua/entities/gmod_wire_ledtape.lua:101]: Also should be using ipairs

File: lua/entities/gmod_wire_ledtape.lua
Comment (Fasteroid) [reply lua/entities/gmod_wire_ledtape.lua:129]: iirc this needs to be done to calculate lighting changes on the base tape layer, but I'll double check.  could probably be optimized a little bit.

File: lua/entities/gmod_wire_ledtape.lua
Comment (Fasteroid) [reply lua/entities/gmod_wire_ledtape.lua:115]: Why?  What about it is problematic?

File: lua/entities/gmod_wire_ledtape.lua
Comment (Fasteroid) [reply lua/entities/gmod_wire_ledtape.lua:99]: I'm not sure what lines this is supposed to point to.

**Comment by Fasteroid** [comment]
If this PR is being blocked for performance issues, I think we have much bigger problems.  This is the current rendering hook for **_normal_** wires:

https://github.com/wiremod/wire/blob/67e3cbc8b7da2ae86b77859ae3f125c1001ca327/lua/weapons/gmod_tool/stools/wire_adv.lua#L1224-L1281

This can be used to drop people's framerates even more than the LED tape does.  We might be looking at rewriting this too for consistency before this PR gets merged.

**Comment by Fasteroid** [comment]
I don't even want to merge this anymore.  Thanks for getting mad at me in DMs for trying to clarify on some of these @Vurv78

**Comment by MorlandMoran** [comment]
> I don't even want to merge this anymore. Thanks for getting mad at me in DMs for trying to clarify on some of these @Vurv78

Wiremod quickly changed to a dictatorship

**Comment by Vurv78** [comment]
> I don't even want to merge this anymore. Thanks for getting mad at me in DMs for trying to clarify on some of these @Vurv78

That's your prerogative. I said I was done with your DMs since all you did was complain about the reviews and I don't want to deal with it atm. You can whine and close the PR, it's yours, so.

File: lua/entities/gmod_wire_ledtape.lua
Comment (Anticept) [reply lua/entities/gmod_wire_ledtape.lua:134]: Consistency, and little things like this make for easier maintenance in the long run.

**Comment by thegrb93** [comment]
It's good to suggest potential performance improvements. Nitpicking them as if that's what it should be, without doing timing testing isn't correct though. I suggest measuring performance with suggestions and taking the best performing code.

File: lua/wire/stools/ledtape.lua
Comment (Fasteroid) [reply lua/wire/stools/ledtape.lua:56]: Still waiting for an answer...

**Comment by Fasteroid** [comment]
> Wiremod quickly changed to a dictatorship

He's not the one who closed it.

**Comment by Anticept** [comment]
I've asked all parties not to bring this kind of energy to github if a personal issue arises. Not everyone will get along, or agree with the process, but it happens. Let's try to avoid using github PRs as the battleground though. Thank you everyone!

**Pull request closed without merging**
